### ISSUE TITLE
Orchard does not work with Kendo grid

### ISSUE THREAD

**Issue by ydenega** [OrchardCMS/Orchard2#768]
If you add a Kendo grid into 
Orchard2-master\src\OrchardCore.Modules\Orchard.Users\Views\Admin\Index.cshtml (and 
 OrderViewModel + corresponding methods to a controller),  Orchard code throws NullReferenceException.
It works OK if you comment out the following likes (all related to grid being Editable):
      columns.Command(command => { command.Edit(); command.Destroy(); }).Width(160);
      .ToolBar(toolbar => toolbar.Create())
      .Editable(editable => editable.Mode(GridEditMode.PopUp))

This is a modified Index.cshtml:

@model UsersIndexViewModel
@using Kendo.Mvc.UI
@using Microsoft.AspNetCore.Routing
@using Orchard.Users.ViewModels;

<h1>@RenderTitleSegments(T["Users"])</h1>

<nav class="admin-toolbar">
    <ul class="navbar-nav mr-auto">
        <li class="nav-item">
            <a asp-route-action="Create" class="btn btn-primary" role="button">@T["Add User"]</a>
        </li>
    </ul>
    <form asp-action="Index" method="get" class="form-inline">
        <input id="search-box" asp-for="Options.Search" class="form-control mr-sm-2" type="text" placeholder="@T["Search"]" autofocus="autofocus">
        <button type="submit" class="btn btn-secondary my-2 my-sm-0">@T["Search"]</button>
    </form>
</nav>

@using (Html.BeginForm())
{
    <ul class="list-group">
        @foreach (var entry in Model.Users)
        {
            <li class="list-group-item">
                <div class="properties">
                    <div class="related">
                        <a asp-action="Edit" asp-route-id="@entry.User.Id" class="btn btn-primary btn-sm">@T["Edit"]</a>
                        <a asp-action="Delete" asp-route-id="@entry.User.Id" class="btn btn-danger btn-sm" itemprop="RemoveUrl UnsafeUrl">@T["Delete"]</a>
                    </div>

                    @entry.User.UserName
                </div>
            </li>
        }
    </ul>
}

@(Html.Kendo().Grid<OrderViewModel>()
      .Name("grid")
      .Columns(columns =>
      {
          columns.Bound(p => p.OrderID).Filterable(false);
          columns.Bound(p => p.Freight);
          columns.Bound(p => p.OrderDate).Format("{0:MM/dd/yyyy}");
          columns.Bound(p => p.ShipName);
          columns.Bound(p => p.ShipCity);
          columns.Command(command => { command.Edit(); command.Destroy(); }).Width(160);
      })
      .ToolBar(toolbar => toolbar.Create())
      .Editable(editable => editable.Mode(GridEditMode.PopUp))
      .Pageable()
      .Sortable()
      .Scrollable()
      .Filterable()
      .HtmlAttributes(new { style = "height:550px;" })
      .DataSource(dataSource => dataSource
          .Ajax()
          .PageSize(20)
          .Read(read => read.Action("Orders_Read", "Admin").Data("sendAntiForgery"))
          .Update(update => update.Action("Orders_Read", "Admin").Data("sendAntiForgery"))
          .Create(create => create.Action("Orders_Read", "Admin").Data("sendAntiForgery"))
          .Destroy(destroy => destroy.Action("Orders_Read", "Admin").Data("sendAntiForgery"))
          .ServerOperation(true)
      )
      )

@await DisplayAsync(Model.Pager)

<script type="text/javascript">
    function sendAntiForgery() {
        return { "__RequestVerificationToken": $('input[name=__RequestVerificationToken]').val() }
    }
</script>

Technically it fails because 
public void PopulateValues(ViewLocationExpanderContext context)
in CompositeViewLocationExpanderProvider
is called with AreaName in the context as null, so the code throws an exception down the stack.

**Comment by sebastienros** [comment]
Have you tried to set the "area" property when you generate the links? It's is required in orchard, and need to be the id of the module.

**Comment by ydenega** [comment]
Actually, I did try that, without success.
I identified two pleces in the code (identified as // Y.D.) , and if I do these small fixes, everything seems to work just fine. Unfortunately I cannot do these fixes in my app as I am using Orchard modules from GitHub. These are the changes that fix the issue:

1. Orchard.Environment.Extensions ExtensionManager:
        public IExtensionInfo GetExtension(string extensionId)
        {
            EnsureInitialized();

            // Y.D.
            if (string.IsNullOrEmpty(extensionId))
            {
                return new NotFoundExtensionInfo(extensionId);
            }

2. Orchard.DisplayManagement ThemeAwareViewLocationExpanderProvider:
        public virtual IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context,
                                                               IEnumerable<string> viewLocations)
        {
            // Y.D.
            if (string.IsNullOrEmpty(context.AreaName))
            {
                return Enumerable.Empty<string>();
            }

There is also a third place that I have changed to make it work with Kendo:
Microsoft.AspNetCore.Mvc.Modules ModularServiceCollectionExtensions

        public static IServiceCollection AddMvcModules(this IServiceCollection services,
            IServiceProvider applicationServices)
        {
            var builder = services.AddMvcCore(options =>
            {
                // Do we need this?
                options.Filters.Add(typeof(AutoValidateAntiforgeryTokenAuthorizationFilter));

                options.ModelBinderProviders.Insert(0, new CheckMarkModelBinderProvider());
            });

            builder.AddAuthorization();
            builder.AddViews();
            builder.AddViewLocalization();

            AddModularFrameworkParts(applicationServices, builder.PartManager);

            builder.AddModularRazorViewEngine(applicationServices);

            // Use a custom ICompilerCacheProvider so all tenant reuse the same ICompilerCache instance
            builder.Services.Replace(new ServiceDescriptor(typeof(ICompilerCacheProvider), typeof(SharedCompilerCacheProvider), ServiceLifetime.Singleton));

            AddMvcModuleCoreServices(services);
            AddDefaultFrameworkParts(builder.PartManager);

            // Order important
            builder.AddJsonFormatters();
            // Y.D.
            builder.AddJsonOptions(options => options.SerializerSettings.ContractResolver = new DefaultContractResolver());

            return services;
        }

Otherwise formatters on the server and client sides are out of sync. However this one has an easy workaround, you can just specify the correct formatter when returning from a Controller's method.

The first two though look like bugs to me. 

Also, if you use JQuery version of Kendo UI (no Razor code), it works. It is a pain however, Razor version is so much easier to work with.

**Comment by ydenega** [comment]
Hi,
Can you please mark it as a bug? This issue is really a big problem to us.

**Comment by sebastienros** [comment]
Can you create a PR? It would be easier to analyze the reasons and push it to the main source.

But for the formatter I think you should add it with from your code or one of your modules. Unless you think we have changed it by mistake, then please tell us where and why it's wrong.

**Comment by ydenega** [comment]
I agree about the formatter. It is not an issue for us as it is easy to override the behavior in our own code. I will create a PR, I thought I needed a bug in order to do it.
Thank you.

**Comment by sebastienros** [comment]
That's a fair point, tagged ;)

**Comment by ydenega** [comment]
Sorry to bother you again, but I am having trouble creating a pull request. I created a branch for my changes, but I cannot push it to GitHub. Looks like I do not have permissions to do it:
C:\Users\ydene\Source\Repos\Orchard2>git push --set-upstream origin KendoUISupport
remote: Permission to OrchardCMS/Orchard2.git denied to ydenega.
fatal: unable to access 'https://github.com/OrchardCMS/Orchard2.git/': The requested URL returned error: 403

Please help me out here. I am new ti this technology.

**Comment by agriffard** [comment]
You need to make your own fork of the repository in your github account, make the changes in this fork and then submit a PR from this fork to Orchard2.

**Comment by ydenega** [comment]
Thank you.
